File: linda-core/src/main/java/it/unibo/coordination/linda/core/InspectableTupleSpace.java
package it.unibo.coordination.linda.core;

import it.unibo.coordination.linda.core.events.OperationEvent;
import it.unibo.coordination.linda.core.events.TupleEvent;
import it.unibo.coordination.utils.events.EventSource;

public interface InspectableTupleSpace<T extends Tuple, TT extends Template> extends TupleSpace<T, TT> {
    EventSource<OperationEvent<T, TT>> operationInvoked();

    EventSource<OperationEvent<T, TT>> operationCompleted();

    EventSource<TupleEvent<T, TT>> tupleSpaceChanged();
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/InspectableTupleSpace.kt
package it.unibo.coordination.linda.core

import it.unibo.coordination.linda.core.traits.Inspectability
import it.unibo.coordination.linda.core.traits.InspectabilityOperationalPrimitives

interface InspectableTupleSpace<T : Tuple<T>, TT : Template<T>, K, V, M : Match<T, TT, K, V>>
    : TupleSpace<T, TT, K, V, M>,
        Inspectability<T, TT>,
        InspectabilityOperationalPrimitives<T, TT>